src/cargo/core/shell.rs

```diff
@@ -119,199 +91,82 @@ impl MultiShell {
+            Some("never") => Choice::Never,
 
-            None => Auto,
+            Some("auto") |
+            None => Choice::CargoAuto,
 
             Some(arg) => bail!("argument for --color must be auto, always, or \
                                 never, but found `{}`", arg),
-        };
-        self.out.set_color_config(cfg);
-        self.err.set_color_config(cfg);
-        Ok(())
-    }
-
-    pub fn get_verbose(&self) -> Verbosity {
-        self.verbosity
+       };
+        self.choice = cfg;
+        self.err = StandardStream::stderr(cfg.to_color_choice());
+        return Ok(());
     }
 
-    pub fn color_config(&self) -> ColorConfig {
-        assert!(self.out.config.color_config == self.err.config.color_config);
-        self.out.config.color_config
+    pub fn color_choice(&self) -> Choice {
+        self.choice
     }
 }
 
-impl Shell {
-    pub fn create<T: FnMut() -> Box<Write + Send>>(mut out_fn: T, config: ShellConfig) -> Shell {
-        let term = match Shell::get_term(out_fn()) {
-            Ok(t) => t,
-            Err(_) => NoColor(out_fn())
-        };
-
-        Shell {
-            terminal: term,
-            config: config,
-        }
-    }
-
-    #[cfg(any(windows))]
-    fn get_term(out: Box<Write + Send>) -> CargoResult<AdequateTerminal> {
-        // Check if the creation of a console will succeed
-        if ::term::WinConsole::new(vec![0u8; 0]).is_ok() {
-            let t = ::term::WinConsole::new(out)?;
-            if !t.supports_color() {
-                Ok(NoColor(Box::new(t)))
-            } else {
-                Ok(Colored(Box::new(t)))
-            }
-        } else {
-            // If we fail to get a windows console, we try to get a `TermInfo` one
-            Ok(Shell::get_terminfo_term(out))
-        }
-    }
-
-    #[cfg(any(unix))]
-    fn get_term(out: Box<Write + Send>) -> CargoResult<AdequateTerminal> {
-        Ok(Shell::get_terminfo_term(out))
-    }
-
-    fn get_terminfo_term(out: Box<Write + Send>) -> AdequateTerminal {
-        // Use `TermInfo::from_env()` and `TerminfoTerminal::supports_color()`
-        // to determine if creation of a TerminfoTerminal is possible regardless
-        // of the tty status. --color options are parsed after Shell creation so
-        // always try to create a terminal that supports color output. Fall back
-        // to a no-color terminal regardless of whether or not a tty is present
-        // and if color output is not possible.
-        match ::term::terminfo::TermInfo::from_env() {
-            Ok(ti) => {
-                let term = TerminfoTerminal::new_with_terminfo(out, ti);
-                if !term.supports_color() {
-                    NoColor(term.into_inner())
-                } else {
-                    // Color output is possible.
-                    Colored(Box::new(term))
-                }
-            },
-            Err(_) => NoColor(out),
+impl Choice {
+    fn to_color_choice(&self) -> ColorChoice {
```

Comment:
I am not an expert in these things, but if we and termcolor do different things then perhaps at least one is wrong :) 

There's also a difference for windows, where the term color in always mode would output ansi escapes if console API is not available, but term is set. If we are happy with this behavior, we can avoid calling winapi functions here and let termcolor do this.